### PR TITLE
Alerting: use folder UID instead of title in rule API

### PR DESCRIPTION
**Pull request by VikaCep**
After https://github.com/grafana/grafana/pull/74600, Grafana rule names will be returned from the API as a combination of the folder name and parent UID separated by `/`. (See proposal 2 from https://docs.google.com/document/d/1tokkNXVWGOmTwdgUoGui3F_sWPdCMBVB2pOn8n_HLdk/edit#heading=h.t7gn8ijy9fiw)

We don't want to show the string portion behind the first `/` in the UI. This PR addresses that. 

Besides that, for POST requests to this API, the frontend needs to send the folder UID (`namespace_uid`) instead of the folder title. 

Before

image

After

image

### CODE REVIEW

File: public/app/features/alerting/unified/api/ruler.ts [hunk 1]
```diff
@@ -36,10 +36,12 @@ export function rulerUrlBuilder(rulerConfig: RulerDataSourceConfig) {
         params: params,
       };
     },
-    namespace: (namespace: string): RulerRequestUrl => ({
-      path: `${rulerPath}/${encodeURIComponent(namespace)}`,
-      params: Object.fromEntries(rulerSearchParams),
-    }),
+    namespace: (namespace: string): RulerRequestUrl => {
```
Comment (soniaAguilarPeiron) [inline public/app/features/alerting/unified/api/ruler.ts:39]: there is no change here, right?

**Comment by soniaAguilarPeiron** [comment]
I'm a little confused with the description about  `rule names`. Same in the code, we are using the term rule name , for example in `formatRuleName`, when what we are actually formatting is the `folder name`. Is that correct?

File: public/app/features/alerting/unified/components/expressions/util.ts [hunk 2]
```diff
@@ -37,10 +40,20 @@ const formatLabels = (labels: Labels): string => {
     .join(', ');
 };
 
+const formatRuleName = (namespace: CombinedRuleNamespace): string => {
```
Comment (soniaAguilarPeiron) [inline public/app/features/alerting/unified/components/expressions/util.ts:43]: Shouldn't this method be named "formatFolderName" instead?

File: public/app/types/unified-alerting-dto.ts [hunk 3]
```diff
@@ -213,7 +213,7 @@ export interface GrafanaRuleDefinition extends PostableGrafanaRuleDefinition {
   id?: string;
   uid: string;
   namespace_uid: string;
-  namespace_id: number;
+  namespace_id?: number; //deprecated
```
Comment (gillesdemey) [inline public/app/types/unified-alerting-dto.ts:216]: You can mark it as deprecated with the following syntax :)
```suggestion
  /** @deprecated use namespace_uid instead */
  namespace_id: number;
```

File: public/app/features/alerting/unified/utils/rulerClient.ts [hunk 4]
```diff
@@ -81,7 +81,11 @@ export function getRulerClient(rulerConfig: RulerDataSourceConfig): RulerClient
   };
 
   const deleteRule = async (ruleWithLocation: RuleWithLocation): Promise<void> => {
-    const { namespace, group, rule } = ruleWithLocation;
+    let { namespace, group, rule } = ruleWithLocation;
+
+    if (isGrafanaRulerRule(rule)) {
```
Comment (gillesdemey) [inline public/app/features/alerting/unified/utils/rulerClient.ts:86]: Instead of overwriting the namespace here, would it make more sense to add the `namespace_uid` / `folder_uid` to the `ruleWithLocation` interface and update the code for [`findEditableRule`](https://github.com/grafana/grafana/blob/577b3f2fb20f189c95d806d5ef5f016f35097933/public/app/features/alerting/unified/utils/rulerClient.ts#L36-L45)?

File: public/app/features/alerting/unified/state/actions.ts [hunk 5]
```diff
@@ -884,7 +895,7 @@ export const updateRulesOrder = createAsyncThunk(
             rules: newRules,
           };
 
-          await setRulerRuleGroup(rulerConfig, namespaceName, payload);
+          await setRulerRuleGroup(rulerConfig, folderUid, payload);
```
Comment (gillesdemey) [inline public/app/features/alerting/unified/state/actions.ts:898]: Should this be `namespaceName ?? folderUid` to make it work for both cloud and GMA rules?

File: public/app/features/alerting/unified/utils/rulerClient.ts (hunk 4 above)
Comment (VikaCep) [reply public/app/features/alerting/unified/utils/rulerClient.ts]: Good idea, done!

File: public/app/features/alerting/unified/components/expressions/util.ts (hunk 2 above)
Comment (VikaCep) [reply public/app/features/alerting/unified/components/expressions/util.ts]: Renamed it 👍

File: public/app/features/alerting/unified/api/ruler.ts (hunk 1 above)
Comment (VikaCep) [reply public/app/features/alerting/unified/api/ruler.ts]: Right, I changed back to how it was before (removing the explicit `return` statement)

**Comment by VikaCep** [comment]
> I'm a bit confused about the description when it mentions rule names. Same in the code, we are using the term rule name , for example in formatRuleName, when what we are actually formatting is the folder name. Is my understanding correct?

Renamed all references to folder name for clarity.

File: public/app/features/alerting/unified/components/expressions/util.ts [hunk 6]
```diff
@@ -37,10 +40,20 @@ const formatLabels = (labels: Labels): string => {
     .join(', ');
 };
 
+const formatFolderName = (namespace: CombinedRuleNamespace): string => {
+  const ruleName = namespace.name;
```
Comment (soniaAguilarPeiron) [inline public/app/features/alerting/unified/components/expressions/util.ts:44]: this is not the rule name, is the folder/namespace name

File: public/app/features/alerting/unified/RuleEditorGrafanaRules.test.tsx [hunk 7]
```diff
@@ -105,29 +105,83 @@ describe('RuleEditor grafana managed rules', () => {
...
+                exec_err_state: GrafanaAlertStateDecision.Error,
+                no_data_state: GrafanaAlertStateDecision.NoData,
+                title: 'my great new rule',
+              },
+            },
+          ],
         },
       ],
     });
+    // mocks.api.fetchRulerRules.mockResolvedValue({
```
Comment (gillesdemey) [inline public/app/features/alerting/unified/RuleEditorGrafanaRules.test.tsx:153]: unused test here? :)

File: public/app/features/alerting/unified/RuleEditorGrafanaRules.test.tsx [hunk 8]
```diff
@@ -142,7 +196,7 @@ describe('RuleEditor grafana managed rules', () => {
 
     await userEvent.type(await ui.inputs.name.find(), 'my great new rule');
 
-    const folderInput = await ui.inputs.folder.find();
+    const folderInput = ui.inputs.folder.get();
```
Comment (gillesdemey) [inline public/app/features/alerting/unified/RuleEditorGrafanaRules.test.tsx:199]: Any particular reason we changed this from `await find()`? I usually see these when a flakey test would otherwise be triggered.

File: public/app/features/alerting/unified/components/expressions/util.ts [hunk 9]
```diff
@@ -37,10 +40,20 @@ const formatLabels = (labels: Labels): string => {
     .join(', ');
 };
 
+const formatFolderName = (namespace: CombinedRuleNamespace): string => {
```
Comment (gillesdemey) [inline public/app/features/alerting/unified/components/expressions/util.ts:43]: nit: prefer to rename this to `decodeGrafanaNamespace` (or similar) with some context around what we're doing here, why we're doing it (nested folder, etc)

File: public/app/features/alerting/unified/RuleEditorGrafanaRules.test.tsx (hunk 8 above)
Comment (soniaAguilarPeiron) [reply public/app/features/alerting/unified/RuleEditorGrafanaRules.test.tsx:199]: good catch, I forgot to restore this change after fixing the test

**Comment by papagian** [comment]
I would suggest merging it in the [backend equivalent](https://github.com/grafana/grafana/pull/74600) (I don't see any harm on that) and continue they review cycle there
it's hard to follow the progress otherwise,  WDYT?

**Comment by soniaAguilarPeiron** [comment]
> I would suggest merging it in the [backend equivalent](https://github.com/grafana/grafana/pull/74600) (I don't see any harm on that) and continue the review cycle there it's hard to follow the progress otherwise, WDYT?

We've identified a scenario where using the `/` character as a separator can cause issues, as certain rules may be provisioned with the `/` character in the folder name. Consequently, we suggest utilizing a non-printable ASCII character, as did in [here](https://github.com/grafana/grafana/blob/0e3c39c0e6d6fb2679b3146549ba5e529574f124/public/app/features/alerting/unified/utils/rule-id.ts#L105-L119).
Once you update these changes in the BE PR, we can update this one using this new separator char.

**Comment by papagian** [comment]
> > I would suggest merging it in the [backend equivalent](https://github.com/grafana/grafana/pull/74600) (I don't see any harm on that) and continue the review cycle there it's hard to follow the progress otherwise, WDYT?
> 
> Sounds good @papagian but we need to address an issue before merging: We've identified a scenario where using the `/` character as a separator can cause issues, as certain rules may be provisioned with the `/` character in the folder name. Consequently, we suggest utilizing a non-printable ASCII character, as did in [here](https://github.com/grafana/grafana/blob/0e3c39c0e6d6fb2679b3146549ba5e529574f124/public/app/features/alerting/unified/utils/rule-id.ts#L105-L119). Once you update these changes in the BE [PR](https://github.com/grafana/grafana/pull/74600), we can update this one using this new separator char.

I've started escaping slashes in folder title in f8784bcb04f2d60a52b097c532a2f8d471cfefc6 so I think that all you have to do is to remove them when showing the folder titles in the UI

**Comment by soniaAguilarPeiron** [comment]
> > > I would suggest merging it in the [backend equivalent](https://github.com/grafana/grafana/pull/74600) (I don't see any harm on that) and continue the review cycle there it's hard to follow the progress otherwise, WDYT?
> > 
> > 
> > Sounds good @papagian but we need to address an issue before merging: We've identified a scenario where using the `/` character as a separator can cause issues, as certain rules may be provisioned with the `/` character in the folder name. Consequently, we suggest utilizing a non-printable ASCII character, as did in [here](https://github.com/grafana/grafana/blob/0e3c39c0e6d6fb2679b3146549ba5e529574f124/public/app/features/alerting/unified/utils/rule-id.ts#L105-L119). Once you update these changes in the BE [PR](https://github.com/grafana/grafana/pull/74600), we can update this one using this new separator char.
> 
> I've started escaping slashes in folder title in [f8784bc](https://github.com/grafana/grafana/commit/f8784bcb04f2d60a52b097c532a2f8d471cfefc6) so I think that all you have to do is to remove them when showing the folder titles in the UI

Nice, I merged from your branch and it's working correctly. I provisioned a folder with a leading slash /FOLDER 5 and everything looks good. @gillesdemey 👀 

Screenshot 2023-11-22 at 14 55 35


Screenshot 2023-11-22 at 14 55 18

**Comment by papagian** [comment]
> Nice, I merged from your branch and it's working correctly. I provisioned a folder with a leading slash /FOLDER 5 and everything looks good. @gillesdemey 👀
> 
Screenshot 2023-11-22 at 14 55 18

FYI rule exporting (and handling properly folders containing slashes is addressed in another [PR](https://github.com/grafana/grafana/pull/77450)

**Comment by soniaAguilarPeiron** [comment]
ok, sorry, I did not test the correct use case (I tested with `\ `instead of` /`). I'm going to test it with the right use cases. @papagian

**Comment by soniaAguilarPeiron** [comment]
> I've started escaping slashes in folder title in [f8784bc](https://github.com/grafana/grafana/commit/f8784bcb04f2d60a52b097c532a2f8d471cfefc6) so I think that all you have to do is to remove them when showing the folder titles in the UI

After this change the UI is not receiving the leading folderuid  + separator anymore. is this expected behaviour? @papagian 

Before escaping slashes commit:

Screenshot 2023-11-23 at 09 46 22

After escaping slashes commit:

Screenshot 2023-11-23 at 09 49 00

**Comment by papagian** [comment]
> > I've started escaping slashes in folder title in [f8784bc](https://github.com/grafana/grafana/commit/f8784bcb04f2d60a52b097c532a2f8d471cfefc6) so I think that all you have to do is to remove them when showing the folder titles in the UI
> 
> After this change the UI is not receiving the leading folderuid + separator anymore. is this expected behaviour? @papagian
> 
> Before escaping slashes commit:
> 
> Screenshot 2023-11-23 at 09 46 22
> After escaping slashes commit:
> 
> Screenshot 2023-11-23 at 09 49 00

yes; so I think the frontend should replace `\/` with `/` so it will become: `/FOLDER/ 5`

**Comment by soniaAguilarPeiron** [comment]
> yes; so I think the frontend should replace \/ with / so it will become: /FOLDER/ 5

   how do we know in the FE if this escaped slash means that `FOLDER` is the folderUID or is a part of the folder title?

**Comment by papagian** [comment]
> > yes; so I think the frontend should replace / with / so it will become: /FOLDER/ 5
> 
> how do we know in the FE if this escaped slash means that `FOLDER` is the folderUID or is a part of the folder title?

the format is: `<parent folder UID>/<folder title>` so if it starts with `/` means the its a root folder (does not live under another folder)

**Review by gillesdemey — APPROVED** [review]
LGTM! 🚀
